File: 0437-path-sum-iii/cpp/Solution.cpp
class Solution {
public:
    int count = 0;
    int pathSum(TreeNode* root, int sum) {
        dfs(root, sum);
        return count;
    }
    unordered_map<int, int> dfs(TreeNode* root, int sum) {
        unordered_map<int, int> map;
        if (!root) return map;
        unordered_map<int, int> leftValues = dfs(root->left, sum);
        unordered_map<int, int> rightValues = dfs(root->right, sum);
        unordered_map<int, int> rootValues = map;
        rootValues.insert({root->val, 1});
        for (auto it = leftValues.begin(); it != map.end(); ++it) {
            int newKey = it->first + root->val;
            int newValue = it->second;
            if (rootValues.find(newKey) != rootValues.end()) {
                newValue += rootValues[newKey];
            }
            rootValues[newKey] = newValue;
        }
        for (auto it = rightValues.begin(); it != rightValues.end(); ++it) {
            int newKey = it->first + root->val;
            int newValue = it->second;
            if (rootValues.find(newKey) != rootValues.end()) {
                newValue += rootValues[newKey];
            }
            rootValues[newKey] = newValue;
        }
        for (auto it = rootValues.begin(); it != rootValues.end(); ++it) {
            if (it->first == sum) {
                count += it->second;
            }
        }
        return rootValues;
    }
};

class Solution {
public:
    int count = 0;
    int pathSum(TreeNode* root, int targetSum) {
        vector<int> path;
        dfs(root, path, targetSum);
        return count;
    }
    void dfs(TreeNode* root, vector<int>& path, int sum) {
        if (root == nullptr) return;
        path.push_back(root->val);
        int curSum = 0;
        for (int i = path.size() - 1; i >= 0; --i) {
            curSum += path[i];
            if (curSum == sum) count++;
        }
        dfs(root->left, path, sum);
        dfs(root->right, path, sum);
        path.pop_back();
    }
};

File: 0437-path-sum-iii/python/Solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self):
        self.count = 0

    def pathSum(self, root: TreeNode, targetSum: int) -> int:
        self.dfs(root, targetSum)
        return self.count

    def dfs(self, root, targetSum):
        if not root:
            return {}
        leftValues = self.dfs(root.left, targetSum)
        rightValues = self.dfs(root.right, targetSum)
        rootValues = dict()
        rootValues[root.val] = 1
        for key, value in leftValues.items():
            key = key + root.val
            if key in rootValues:
                value += rootValues[key]
            rootValues[key] = value
        for key, value in rightValues.items():
            key = key + root.val
            if key in rootValues:
                value += rootValues[key]
            rootValues[key] = value
        for key, value in rootValues.items():
            if key == targetSum:
                self.count += value
        return rootValues


class Solution:
    def __init__(self):
        self.count = 0
    def pathSum(self, root: TreeNode, targetSum: int) -> int:
        self.dfs(root,[], targetSum)
        return self.count
    def dfs(self, root, path, targetSum):
        if not root:
            return
        path.append(root.val)
        curSum = 0
        for i in range(len(path)-1, -1,-1):
            curSum += path[i]
            if curSum == targetSum:
                self.count += 1
        self.dfs(root.left, path, targetSum)
        self.dfs(root.right, path, targetSum)
        path.pop()